templates/tree-editor/tree/tree-schema.ts

```diff
@@ -1,101 +1,389 @@
+            {
+              'type': 'Control',
+              'label': 'Length',
+              'scope': '#/properties/dimension/properties/length'
+            }
+          ]
+        }
+      ]
+    },
+    {
+      'type': 'Control',
+      'label': 'Ram',
+      'scope': '#/properties/ram'
+    },
+    {
+      'type': 'Group',
+      'label': 'User Description',
+      'elements': [
         {
           'type': 'Control',
-          'label': 'Last Name',
-          'scope': '#/properties/person/properties/lastname'
+          'label': 'User Description',
+          'scope': '#/properties/userDescription'
         }
       ]
     }
   ]
 };
 
-export const componentView = {
+export const machineView = {
   'type': 'VerticalLayout',
   'elements': [
     {
-      'type': 'Control',
-      'label': 'Label',
-      'scope': '#/properties/label'
+      'type': 'Label',
+      'text': 'Machine'
     },
     {
-      "type": "VerticalLayout",
-      "elements": [
-        {
-          "type": "Control",
-          "scope": "#/properties/persons"
-        }
-      ]
+      'type': 'Control',
+      'label': 'Name',
+      'scope': '#/properties/name'
     }
   ]
 };
 
-export const personView = {
+export const brewingView = {
+  'type': 'VerticalLayout',
+  'elements': [
+    {
+      'type': 'Label',
+      'text': 'Brewing Unit'
+    },
+    {
+      'type': 'Control',
+      'label': 'Temperature (°C)',
+      'scope': '#/properties/temperature'
+    }
+  ]
+};
+export const dripTrayView = {
   'type': 'VerticalLayout',
   'elements': [
+    {
+      'type': 'Label',
+      'text': 'DripTray'
+    },
     {
       'type': 'Control',
-      'label': 'First Name',
-      'scope': '#/properties/firstname'
+      'label': 'Material',
+      'scope': '#/properties/material'
+    }
+  ]
+};
+
+export const waterTankView = {
+  'type': 'VerticalLayout',
+  'elements': [
+    {
+      'type': 'Label',
+      'text': 'Water Tank'
     },
     {
       'type': 'Control',
-      'label': 'Last Name',
-      'scope': '#/properties/lastname'
+      'label': 'Capacity (ml)',
+      'scope': '#/properties/capacity'
     }
   ]
-}
+};
 
-export const schema = {
+export const coffeeSchema = {
   'definitions': {
-    'component': {
-      'title': 'Component',
+    'machine': {
+      '$id': '#machine',
+      'title': 'Machine',
       'properties': {
         'typeId': {
-          'const': '#component'
+          'const': 'Machine'
         },
-        'label': {
+        'name': {
           'type': 'string'
         },
-        'persons': {
+        'children': {
           'type': 'array',
           'items': {
-            '$ref': '#/definitions/person'
+            'anyOf': [
+              { '$ref': '#/definitions/controlunit' },
+              { '$ref': '#/definitions/brewingunit' },
+              { '$ref': '#/definitions/driptray' },
+              { '$ref': '#/definitions/watertank' }
+            ]
           }
-        },
+        }
       },
```

Comment:
The children should be removed as we manually code them anyway.

Reply:
Yes but shouldn't the schema still define all properties of a type? Otherwise data created in the editor wouldn't even validate against its own schema - this seems a little strange to me. But maybe I'm missing a problem with this?

Reply:
The model is validated (in JSON Forms) but the result of this `children` validation isn't shown anywhere. So we have no benefit by adding it here (besides that this relationship is modeled but ignored in the code).
The downsides are:
 - The user might expect modeling children here is necessary or changes the behavior of the tree while it has no effect
 - While there is no visual effect, JSON Forms will actually still resolve the references. There is a known issue of circular dependencies taking a long time to resolve. So should the user add circular references here they will bring down the performance.

Reply:
Ok, that makes sense, thanks for the explanation.